extensions/graphql/src/__tests__/integration/export-graphql-spec.integration.ts

```diff
@@ -0,0 +1,114 @@
+// Copyright IBM Corp. 2021. All Rights Reserved.
+// Node module: @loopback/graphql
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {BootMixin} from '@loopback/boot';
+import {ApplicationConfig} from '@loopback/core';
+import {RestApplication} from '@loopback/rest';
+import {expect, TestSandbox} from '@loopback/testlab';
+import fs from 'fs';
+import path from 'path';
+import {format} from 'util';
+import {GraphQLBindings, GraphQLComponent, GraphQLServer} from '../..';
+import {RecipeResolver} from '../fixtures/graphql-resolvers/recipe/recipe.resolver';
+
+const sandbox = new TestSandbox(path.resolve(__dirname, '../../../.sandbox'));
+
+const EXPECTED_SPEC = `schema {
+  query: Query
+}
+
+type Query {
+  recipes: [Recipe!]!
+}
+
+"""Object representing cooking recipe"""
+type Recipe {
+  id: ID!
+  title: String!
+}
+`;
+
+describe('exportGraphQLSchema', () => {
+  describe('standalone GraphQLServer', () => {
+    let server: GraphQLServer;
+    let lastLog: string;
+
+    const log = (formatter: unknown, ...args: unknown[]) => {
+      lastLog = format(formatter, ...args);
+    };
+
+    beforeEach(async () => {
+      givenServer();
+      server.resolver(RecipeResolver);
+    });
+    function givenServer() {
+      server = new GraphQLServer();
+    }
+    it('renders the expected spec to stdout', async () => {
+      await server.exportGraphQLSchema('-', log);
+      expect(lastLog).to.eql(EXPECTED_SPEC);
+    });
+    it('renders the expected spec to file', async () => {
+      const file = path.join(sandbox.path, 'standalone.spec.gql');
+      await server.exportGraphQLSchema(file);
+      const content = fs.readFileSync(file, 'utf-8');
+      expect(content).to.eql(EXPECTED_SPEC);
+    });
+  });
+
+  describe('Booted RestApplication with GraphQL as middleware', () => {
+    let app: TestApplication;
+    let lastLog: string;
+
+    const log = (formatter: unknown, ...args: unknown[]) => {
+      lastLog = format(formatter, ...args);
+    };
+    beforeEach(async () => {
+      await givenApp();
+    });
+
+    it('renders to stdout', async () => {
+      await app.exportGQLFile();
+      expect(lastLog).to.eql(EXPECTED_SPEC);
+    });
+
+    it('renders the expected spec to file', async () => {
+      const file = path.join(sandbox.path, 'standalone.spec.gql');
```

Comment:
Currently in the extension, we can generate a file from the `buildSchema` method, however for some reason it is commented out. The file generated is named `schema.gql`. As far as I understand this PR is aimed to provide the same functionality but with custom directives, hence the use for the custom print from the `graphql-tools` package. If that's correct then we should name the generated file something like `schema-cd.gql` or something similar.

Reply:
Actually, this PR aims to allow the creation of the spec file via a command, without requiring the server to be run.  The commented code occurs during server startup.

It requires the user to either specify a file name themselves, or to redirect `stdout` to the location of their choice.